client/src/reducers/controls.js

```diff
@@ -100,6 +100,28 @@ function selectedValuesForCategory(categorySelectionState) {
   return selectedValues;
 }
 
+/*
+build a crossfilter dimension map for all gene expression related dimensions.
+*/
+function createGenesDimMap(userDefinedGenes, diffexpGenes, world, crossfilter) {
+  function _createGenesDimMap(genes, nameF) {
+    return genes.reduce((acc, gene) => {
+      acc[nameF(gene)] = World.createVarDimension(
+        world,
+        world.varDataCache,
+        crossfilter,
+        gene
+      );
+      return acc;
+    }, {});
+  }
+
+  return {
+    ..._createGenesDimMap(userDefinedGenes, userDefinedDimensionName),
+    ..._createGenesDimMap(diffexpGenes, diffexpDimensionName)
```

Comment:
I'm don't understand what the spread operator is doing in the case.

Reply:
this is classic reducer - we need to return a new object, but we want it populated with the contents of these other objects.  So the spread is an object literal spread: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax#Spread_in_object_literals

Reply:
I phrased this poorly. I know what the spread operator does, but I'm used to seeing it in reducers applied to state. I think what I actually don't understand is how are dimension maps being used in the controls reducer. Once I understand that I can better review the dimmap changes in the reducer.

Reply:
In this case, it is setting state (or part of the state).   The reason I used a reducer here is for clarity - without it, I would have to construct the function return value (an object) by using Object.assign().  Semantically equivalent, but our coding style prefers the clarity of spread operators.

Reply:
setting "part of the state" is what I was missing. thanks